File: src/modules/limits/limit.service.ts
import { PrismaClient, LimitType } from '@prisma/client';
import { v4 as uuidv4 } from 'uuid';
import { NotFoundError, BusinessRuleError } from '../../shared/errors';

export interface SpendingLimitInput {
  limitType: LimitType;
  valueCents?: number | null;
  mccCode?: string;
}

export class LimitService {
  constructor(private readonly db: PrismaClient) {}

  async setAccountLimits(accountId: string, limits: SpendingLimitInput[]) {
    const account = await this.db.account.findUnique({
      where: { id: accountId },
      select: { id: true, creditLimitCents: true },
    });
    if (!account) throw new NotFoundError('Account', accountId);

    this.validateLimitsAgainstCreditLimit(limits, account.creditLimitCents);

    await this.db.$transaction(async (tx) => {
      for (const limit of limits) {
        const existing = await tx.accountSpendingLimit.findFirst({
          where: { accountId, limitType: limit.limitType, mccCode: limit.mccCode ?? null },
          select: { id: true },
        });
        if (existing) {
          await tx.accountSpendingLimit.update({
            where: { id: existing.id },
            data: { valueCents: limit.valueCents ?? null },
          });
        } else {
          await tx.accountSpendingLimit.create({
            data: {
              id: uuidv4(),
              accountId,
              limitType: limit.limitType,
              valueCents: limit.valueCents ?? null,
              mccCode: limit.mccCode ?? null,
            },
          });
        }

      }
    });

    return this.getAccountLimits(accountId);
  }

  async getAccountLimits(accountId: string) {
    const account = await this.db.account.findUnique({
      where: { id: accountId },
      select: { id: true },
    });
    if (!account) throw new NotFoundError('Account', accountId);

    const limits = await this.db.accountSpendingLimit.findMany({
      where: { accountId },
      orderBy: { createdAt: 'asc' },
    });

    return { accountId, limits };
  }

  async setCardLimits(cardId: string, limits: SpendingLimitInput[]) {
    const card = await this.db.card.findUnique({
      where: { id: cardId },
      select: { id: true, accountId: true },
    });
    if (!card) throw new NotFoundError('Card', cardId);

    const account = await this.db.account.findUnique({
      where: { id: card.accountId },
      select: { creditLimitCents: true },
    });
    if (!account) throw new NotFoundError('Account', card.accountId);

    this.validateLimitsAgainstCreditLimit(limits, account.creditLimitCents);

    await this.db.$transaction(async (tx) => {
      for (const limit of limits) {
        const existingCard = await tx.cardSpendingLimit.findFirst({
          where: { cardId, limitType: limit.limitType, mccCode: limit.mccCode ?? null },
          select: { id: true },
        });
        if (existingCard) {
          await tx.cardSpendingLimit.update({
            where: { id: existingCard.id },
            data: { valueCents: limit.valueCents ?? null },
          });
        } else {
          await tx.cardSpendingLimit.create({
            data: {
              id: uuidv4(),
              cardId,
              limitType: limit.limitType,
              valueCents: limit.valueCents ?? null,
              mccCode: limit.mccCode ?? null,
            },
          });
        }
      }
    });

    return this.getCardLimits(cardId);
  }

  async getCardLimits(cardId: string) {
    const card = await this.db.card.findUnique({
      where: { id: cardId },
      select: { id: true },
    });
    if (!card) throw new NotFoundError('Card', cardId);

    const limits = await this.db.cardSpendingLimit.findMany({
      where: { cardId },
      orderBy: { createdAt: 'asc' },
    });

    return { cardId, limits };
  }

  private validateLimitsAgainstCreditLimit(
    limits: SpendingLimitInput[],
    creditLimitCents: number,
  ) {
    for (const limit of limits) {
      if (
        limit.valueCents != null &&
        ([LimitType.DAILY, LimitType.MONTHLY, LimitType.PER_TRANSACTION] as LimitType[]).includes(limit.limitType) &&
        limit.valueCents > creditLimitCents
      ) {
        throw new BusinessRuleError(
          'LIMIT_EXCEEDS_CREDIT_LIMIT',
          `${limit.limitType} limit (${limit.valueCents}) exceeds account credit limit (${creditLimitCents})`,
        );
      }
    }

    // Per-transaction must not exceed daily (when both are present in the same batch)
    const daily = limits.find(l => l.limitType === LimitType.DAILY)?.valueCents;
    const txn   = limits.find(l => l.limitType === LimitType.PER_TRANSACTION)?.valueCents;
    if (daily != null && txn != null && txn > daily) {
      throw new BusinessRuleError(
        'TRANSACTION_LIMIT_EXCEEDS_DAILY_LIMIT',
        `Per-transaction limit (${txn}) cannot exceed daily limit (${daily})`,
      );
    }
  }
}

File: spring/src/main/java/com/cam/service/LimitService.java
package com.cam.service;

import com.cam.dto.SpendingLimitItem;
import com.cam.entity.Account;
import com.cam.entity.AccountSpendingLimit;
import com.cam.entity.Card;
import com.cam.entity.CardSpendingLimit;
import com.cam.enums.LimitType;
import com.cam.exception.BusinessRuleException;
import com.cam.exception.NotFoundException;
import com.cam.repository.AccountRepository;
import com.cam.repository.AccountSpendingLimitRepository;
import com.cam.repository.CardRepository;
import com.cam.repository.CardSpendingLimitRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.*;

@Slf4j
@Service
@RequiredArgsConstructor
public class LimitService {

    private static final Set<LimitType> MONETARY_LIMIT_TYPES = Set.of(
        LimitType.DAILY, LimitType.MONTHLY, LimitType.PER_TRANSACTION
    );

    private final AccountRepository accountRepository;
    private final CardRepository cardRepository;
    private final AccountSpendingLimitRepository accountSpendingLimitRepository;
    private final CardSpendingLimitRepository cardSpendingLimitRepository;

    @Transactional
    public Map<String, Object> setAccountLimits(UUID accountId, List<SpendingLimitItem> limits) {
        Account account = accountRepository.findById(accountId)
            .orElseThrow(() -> new NotFoundException("Account", accountId));

        validateLimitsAgainstCreditLimit(limits, account.getCreditLimitCents());

        for (SpendingLimitItem limitItem : limits) {
            Optional<AccountSpendingLimit> existing = accountSpendingLimitRepository
                .findByAccountIdAndLimitTypeAndMccCode(
                    accountId,
                    limitItem.getLimitType(),
                    limitItem.getMccCode()
                );

            if (existing.isPresent()) {
                AccountSpendingLimit limit = existing.get();
                limit.setValueCents(limitItem.getValueCents());
                accountSpendingLimitRepository.save(limit);
            } else {
                AccountSpendingLimit limit = AccountSpendingLimit.builder()
                    .id(UUID.randomUUID())
                    .account(account)
                    .limitType(limitItem.getLimitType())
                    .valueCents(limitItem.getValueCents())
                    .mccCode(limitItem.getMccCode())
                    .build();
                accountSpendingLimitRepository.save(limit);
            }
        }

        return getAccountLimits(accountId);
    }

    @Transactional(readOnly = true)
    public Map<String, Object> getAccountLimits(UUID accountId) {
        accountRepository.findById(accountId)
            .orElseThrow(() -> new NotFoundException("Account", accountId));

        List<AccountSpendingLimit> limits =
            accountSpendingLimitRepository.findByAccountIdOrderByCreatedAtAsc(accountId);

        return Map.of("accountId", accountId, "limits", limits);
    }

    @Transactional
    public Map<String, Object> setCardLimits(UUID cardId, List<SpendingLimitItem> limits) {
        Card card = cardRepository.findById(cardId)
            .orElseThrow(() -> new NotFoundException("Card", cardId));

        Account account = accountRepository.findById(card.getAccount().getId())
            .orElseThrow(() -> new NotFoundException("Account", card.getAccount().getId()));

        validateLimitsAgainstCreditLimit(limits, account.getCreditLimitCents());

        for (SpendingLimitItem limitItem : limits) {
            Optional<CardSpendingLimit> existing = cardSpendingLimitRepository
                .findByCardIdAndLimitTypeAndMccCode(
                    cardId,
                    limitItem.getLimitType(),
                    limitItem.getMccCode()
                );

            if (existing.isPresent()) {
                CardSpendingLimit limit = existing.get();
                limit.setValueCents(limitItem.getValueCents());
                cardSpendingLimitRepository.save(limit);
            } else {
                CardSpendingLimit limit = CardSpendingLimit.builder()
                    .id(UUID.randomUUID())
                    .card(card)
                    .limitType(limitItem.getLimitType())
                    .valueCents(limitItem.getValueCents())
                    .mccCode(limitItem.getMccCode())
                    .build();
                cardSpendingLimitRepository.save(limit);
            }
        }

        return getCardLimits(cardId);
    }

    @Transactional(readOnly = true)
    public Map<String, Object> getCardLimits(UUID cardId) {
        cardRepository.findById(cardId)
            .orElseThrow(() -> new NotFoundException("Card", cardId));

        List<CardSpendingLimit> limits =
            cardSpendingLimitRepository.findByCardIdOrderByCreatedAtAsc(cardId);

        return Map.of("cardId", cardId, "limits", limits);
    }

    private void validateLimitsAgainstCreditLimit(List<SpendingLimitItem> limits, long creditLimitCents) {
        for (SpendingLimitItem limit : limits) {
            if (limit.getValueCents() != null
                && MONETARY_LIMIT_TYPES.contains(limit.getLimitType())
                && limit.getValueCents() > creditLimitCents) {
                throw new BusinessRuleException(
                    "LIMIT_EXCEEDS_CREDIT_LIMIT",
                    limit.getLimitType() + " limit (" + limit.getValueCents()
                        + ") exceeds account credit limit (" + creditLimitCents + ")"
                );
            }
        }

        // Per-transaction must not exceed daily when both are present in the same batch
        Long daily = limits.stream()
            .filter(l -> l.getLimitType() == LimitType.DAILY)
            .map(SpendingLimitItem::getValueCents)
            .filter(Objects::nonNull)
            .findFirst()
            .orElse(null);

        Long txn = limits.stream()
            .filter(l -> l.getLimitType() == LimitType.PER_TRANSACTION)
            .map(SpendingLimitItem::getValueCents)
            .filter(Objects::nonNull)
            .findFirst()
            .orElse(null);

        if (daily != null && txn != null && txn > daily) {
            throw new BusinessRuleException(
                "TRANSACTION_LIMIT_EXCEEDS_DAILY_LIMIT",
                "Per-transaction limit (" + txn + ") cannot exceed daily limit (" + daily + ")"
            );
        }
    }
}
